pkg/config/structure/unmarshal.go

```diff
@@ -0,0 +1,368 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package structure defines a helper to retrieve structured data from the config
+package structure
+
+import (
+	"fmt"
+	"reflect"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/DataDog/datadog-agent/pkg/config/model"
+)
+
+// UnmarshalKey retrieves data from the config at the given key and deserializes it
+// to be stored on the target struct. It is implemented entirely using reflection, and
+// does not depend upon details of the data model of the config.
+// Target struct can use of struct tag of "yaml", "json", or "mapstructure" to rename fields
+func UnmarshalKey(cfg model.Reader, key string, target interface{}) error {
+	source := newNode(reflect.ValueOf(cfg.Get(key)))
+	outValue := reflect.ValueOf(target)
+	if outValue.Kind() == reflect.Pointer {
+		outValue = reflect.Indirect(outValue)
+	}
+	if outValue.Kind() == reflect.Map {
+		return copyMap(outValue, source)
+	}
+	if outValue.Kind() == reflect.Struct {
+		return copyStruct(outValue, source)
+	}
+	if outValue.Kind() == reflect.Slice {
+		return copyList(outValue, source.AsList())
+	}
+	return fmt.Errorf("can only UnmarshalKey to struct or slice, got %v", outValue.Kind())
+}
+
+type nodeType int
+
+const (
+	invalidNodeType nodeType = iota
+	scalarNodeType
+	listNodeType
+	mapNodeType
+	structNodeType
+)
+
+type leafimpl struct {
+	val reflect.Value
+}
+
+type leaf interface {
+	GetBool() (bool, error)
+	GetInt() (int, error)
+	GetString() (string, error)
+}
+
+type nodeListImpl struct {
+	val reflect.Value
+}
+
+type nodeList interface {
+	Size() int
+	Index(int) node
+}
+
+type nodeimpl struct {
+	typ nodeType
+	val reflect.Value
+}
+
+type node interface {
+	GetChild(string) node
+	ChildrenKeys() []string
+	AsList() nodeList
+	AsScalar() leaf
+}
+
+func newNode(v reflect.Value) node {
+	if v.Kind() == reflect.Struct {
+		return &nodeimpl{typ: structNodeType, val: v}
+	}
+	if v.Kind() == reflect.Map {
+		return &nodeimpl{typ: mapNodeType, val: v}
+	}
+	if v.Kind() == reflect.Slice {
+		return &nodeimpl{typ: listNodeType, val: v}
+	}
+	if isScalar(v) {
+		return &nodeimpl{typ: scalarNodeType, val: v}
+	}
+	panic(fmt.Errorf("could not create node from: %v of type %T and kind %v", v, v, v.Kind()))
```

Comment:
Do we want to panic? Should we return `(node, error)` and handle the case of invalid node type?

Reply:
Removed all panics from the file, returning errors everywhere now.